### PR TITLE
Cleanup testdevs on failure

### PR DESCRIPTION
**Pull request by trgill**
Update tests so they can be run multiple times in any order.  Previously stale data on real devices
would cause the some of the loopback tests to fail.  There was also a problem with loopback devices
not not going away after a test completed.

The PR wraps both loopback and real devices in a struct that implements the Drop trait.  When the
devices go out of scope, the clean up code in drop() is invoked.

Added a "flags" parameter to unmount.  These changes pass the "-d" to detach loop on unmount.
The parameter may also be useful in the future for lazy unmounts.

### CODE REVIEW

File: tests/loopbacked_tests.rs [hunk 1]
```diff
@@ -33,19 +33,53 @@ use util::simple_tests::test_setup;
 use util::simple_tests::test_teardown;
 use util::simple_tests::test_thinpool_device;
 
+pub struct LoopTestDev {
+    ld: LoopDevice,
+}
+
+impl LoopTestDev {
+    pub fn new(lc: &LoopControl, path: &std::path::PathBuf) -> LoopTestDev {
```
Comment (mulkieran) [inline tests/loopbacked_tests.rs]: Make type of path ```AsRef<Path>```, see https://github.com/mulkieran/loopdev/pull/1/commits/24026db7d7fc7f242f14783eb007c5b28e393017 for explanation.

File: tests/loopbacked_tests.rs [hunk 2]
```diff
@@ -33,19 +33,53 @@ use util::simple_tests::test_setup;
...
+        // Wipe 1 MiB at the beginning, as data sits around on the files.
+        OpenOptions::new()
+            .read(true)
+            .write(true)
+            .open(path)
+            .unwrap();
+
+        let ld = lc.next_free().unwrap();
+        ld.attach(path.clone(), 0).unwrap();
```
Comment (mulkieran) [inline tests/loopbacked_tests.rs]: I think ```clone``` must be unnecessary here.

File: tests/loopbacked_tests.rs [hunk 3]
```diff
@@ -33,19 +33,53 @@ use util::simple_tests::test_setup;
...
+    }
+
+    pub fn detach(&self) -> Result<()> {
+        self.ld.detach()
+    }
+}
+
+impl Drop for LoopTestDev {
+    fn drop(&mut self) {
+        let _ = self.ld.detach();
```
Comment (mulkieran) [inline tests/loopbacked_tests.rs]: Why don't you ```unwrap()``` here?

File: tests/loopbacked_tests.rs [hunk 4]
```diff
@@ -33,19 +33,53 @@ use util::simple_tests::test_setup;
...
+fn get_devices(count: u8, dir: &TempDir) -> Vec<LoopTestDev> {
     let lc = LoopControl::open().unwrap();
     let mut loop_devices = Vec::new();
 
-    let length = Bytes(IEC::Gi as u64);
-    let wipe_length = Bytes(IEC::Mi as u64).sectors();
     for index in 0..count {
-        let subdir = TempDir::new_in(dir, &index.to_string()).unwrap();
-        let path = subdir.path().join("store");
+        let path = dir.path().join(format!("store{}", &index));
```
Comment (mulkieran) [inline tests/loopbacked_tests.rs]: Sticking with each file in separate subdirectory seems better to me. I believe TempDir cleans it up more nicely.

Comment (trgill) [inline tests/loopbacked_tests.rs]: They are already in a TempDir, right?  You are thinking <TempDir>/<TempDir>/store is better?  I was thinking the 2nd level wasn't needed.

File: src/engine/strat_engine/filesystem.rs [hunk 5]
```diff
@@ -147,10 +147,15 @@ pub fn mount_fs(dev_path: &Path, mount_point: &Path) -> EngineResult<()> {
     Ok(())
 }
 
-pub fn unmount_fs(mount_point: &Path) -> EngineResult<()> {
+pub fn unmount_fs(mount_point: &Path, flags: &str) -> EngineResult<()> {
```
Comment (trgill) [inline src/engine/strat_engine/filesystem.rs]: My first atttempt used:

 let output = try!(Command::new("umount").arg(mount_point).args(flags).output());

But, if the user passed in "" - the umount command failed.

Comment (mulkieran) [inline src/engine/strat_engine/filesystem.rs]: The type you want for ```args``` is ```I``` ```where I: IntoIterator<Item=S>, S: AsRef<OsStr>```. ```str``` does not qualify, but Vec definitely does.

File: tests/loopbacked_tests.rs (hunk 4 above)
Comment (mulkieran) [inline tests/loopbacked_tests.rs]: I think I had my reasons, but it's quite possible that ```<tempdir>/0/store``` is no better than ```<tempdir>/store0```? Feel free to ignore this one.

File: tests/loopbacked_tests.rs (hunk 1 above)
Comment (trgill) [inline tests/loopbacked_tests.rs]: #376

File: tests/loopbacked_tests.rs (hunk 2 above)
Comment (trgill) [inline tests/loopbacked_tests.rs]: ok

File: tests/loopbacked_tests.rs (hunk 3 above)
Comment (trgill) [inline tests/loopbacked_tests.rs]: ok

File: tests/loopbacked_tests.rs [hunk 6]
```diff
@@ -33,19 +33,53 @@ use util::simple_tests::test_setup;
...
+    }
+
+    pub fn detach(&self) -> Result<()> {
+        self.ld.detach()
+    }
+}
+
+impl Drop for LoopTestDev {
+    fn drop(&mut self) {
+        let _ = self.ld.detach().unwrap();
```
Comment (mulkieran) [inline tests/loopbacked_tests.rs]: Leave out the ```_```.

File: tests/loopbacked_tests.rs [hunk 7]
```diff
@@ -85,24 +109,15 @@ fn test_with_spec<F>(count: u8, test: F) -> ()
...
     init_logger();
     let tmpdir = TempDir::new("stratis").unwrap();
-    let loop_devices: Vec<LoopDevice> = get_devices(count, &tmpdir);
+    let loop_devices: Vec<LoopTestDev> = get_devices(count, &tmpdir);
     let device_paths: Vec<PathBuf> = loop_devices
         .iter()
-        .map(|x| x.get_path().unwrap())
-        .collect();
+        .map(|x| x.get_path())
+        .collect::<Vec<PathBuf>>();
```
Comment (mulkieran) [inline tests/loopbacked_tests.rs]: This type annotation ought not to be necessary, since the one above should do the trick.

File: tests/loopbacked_tests.rs [hunk 8]
```diff
@@ -33,19 +33,53 @@ use util::simple_tests::test_setup;
...
+
+        LoopTestDev { ld: ld }
+    }
+
+    fn get_path(&self) -> PathBuf {
+        self.ld.get_path().unwrap()
+    }
+
+    pub fn detach(&self) -> Result<()> {
+        self.ld.detach()
```
Comment (mulkieran) [inline tests/loopbacked_tests.rs]: To be consistant, you might as well unwrap() in the definition of the method, and then not unwrap in its use.

File: tests/loopbacked_tests.rs [hunk 9]
```diff
@@ -33,19 +33,53 @@ use util::simple_tests::test_setup;
...
+            .write(true)
+            .open(path)
+            .unwrap();
+
+        let ld = lc.next_free().unwrap();
+        ld.attach(path, 0).unwrap();
+        // Wipe 1 MiB at the beginning, as data sits around on the files.
+        wipe_sectors(&ld.get_path().unwrap(),
+                     Sectors(0),
+                     Bytes(IEC::Mi as u64).sectors())
```
Comment (mulkieran) [inline tests/loopbacked_tests.rs]: Just ```Bytes(IEC::Mi).sectors()```.

File: tests/loopbacked_tests.rs [hunk 10]
```diff
@@ -33,19 +33,53 @@ use util::simple_tests::test_setup;
 use util::simple_tests::test_teardown;
 use util::simple_tests::test_thinpool_device;
 
+pub struct LoopTestDev {
+    ld: LoopDevice,
+}
+
+impl LoopTestDev {
+    pub fn new(lc: &LoopControl, path: &PathBuf) -> LoopTestDev {
```
Comment (mulkieran) [inline tests/loopbacked_tests.rs]: ```&PathBuf``` -> ```&Path```

File: tests/real_tests.rs [hunk 11]
```diff
@@ -31,10 +31,36 @@ use util::simple_tests::test_setup;
...
+    path: PathBuf,
+}
+
+impl RealTestDev {
+    /// Construct a new test device
+    pub fn new(path: &str) -> RealTestDev {
+        let dev_path = PathBuf::from(path);
+        wipe_sectors(dev_path.as_path(),
+                     Sectors(0),
+                     Bytes(IEC::Mi as u64).sectors())
```
Comment (mulkieran) [inline tests/real_tests.rs]: No need for ```as``` here.

File: tests/real_tests.rs [hunk 12]
```diff
@@ -31,10 +31,36 @@ use util::simple_tests::test_setup;
...
+    }
+
+    fn as_path(&self) -> &Path {
+        &self.path.as_path()
+    }
+}
+
+impl Drop for RealTestDev {
+    fn drop(&mut self) {
+        wipe_sectors(self.as_path(), Sectors(0), Bytes(IEC::Mi as u64).sectors()).unwrap();
```
Comment (mulkieran) [inline tests/real_tests.rs]: Skip ```as u64``` here, too.

**Comment by mulkieran** [comment]
@agrover review?

**Comment by agrover** [comment]
looks good! 👍 

I'm final reviewer, so merging.
